Fix part of #8015 : Domain object for topic-viewer-backend-api.service.ts

Vinit-Dantkale opened the pull request:
## Explanation
Attempt to fix #8015 by building class `TopicDataObjectFactory.ts` and test file with changes made where the service is being used. 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
Topic Viewer Domain Object

Commit:
Changes where service used

Vinit-Dantkale commented:
I feel there's room for improvement with files in PR, request to reviewer from #8015 for advice. Thank You.

kevinlee12 commented:
@Vinit-Dantkale please fix the automated tests.

Commit:
Adding to whitelist

Commit:
Alphabetical order

Vinit-Dantkale commented:
Hi @kevinlee12 , fixed frontend-tests locally but cant find the part which is causing error in Travis and e2e test from my file.

kevinlee12 reviewed (approved):
lgtm for code owner files

Vinit-Dantkale commented:
Thank You @kevinlee12.

marianadasilvadev commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
```
Change the year to 2020 only in files you've created in this PR.

Vinit-Dantkale replied:
Oh right, didn't notice it. Thank You @marianazangrossi.

marianadasilvadev reviewed (approved):
LGTM, just a minor nit

Commit:
Date change in files

DubeySandeep commented:
@aks681, Can you please review this PR?

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+ * objects.
+ */
+
+export class TopicData {
```
I would suggest renaming this to `RepresentableTopic` or `ViewableTopic` or `DisplayableTopic` or something around that. 

In general don't end the domain object name with `Data` as data is very generic and it doesn't explains what the domain object contain, like in this case what part of topic data it contains.

@aks681, do you have any suggestion on this?

Vinit-Dantkale replied:
I think `ViewableTopic` sounds good, though can replace as per choice.

aks681 replied:
This is for the viewer page, right? I would suggest `ReadOnlyTopic`, going by the other objects that are named similarly (ReadOnlyStoryNode and ReadOnlySubtopicPage).

Vinit-Dantkale replied:
Using `ReadOnlyTopic` instead of `TopicData`

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+export class TopicData {
+  _topicName: string;
+  _topicId: string;
+  _canonicalStories: any;
```
Why is this `any`?

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+  _additionalStories: any;
+  _uncategorizedSkillIds: Array<string>;
+  _subtopics: any;
+  _degreesOfMastery: any;
```
Ditto here and elsewhere.

In general, try not to use `any` keyword. If you have to use in worst case then try adding a comment explaining the reason for using `any`.

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactorySpec.ts:
```diff
@@ -0,0 +1,99 @@
+  'domain/topic_viewer/TopicDataObjectFactory';
+
+describe('Topic Data Object Factory', () => {
+  let topicDataObjectFactory: TopicDataObjectFactory;
```
Do we have to initialize this with `null`?

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactorySpec.ts:
```diff
@@ -0,0 +1,99 @@
+  });
+
+  it('should get the values as expected', () => {
+    expect(_sampleTopicData.getTopicName()).toEqual('topic_name');
```
Create one test for each function. In general one `it` block should test one behaviour.

Vinit-Dantkale replied:
Are you saying for each function in `TopicData` class?

DubeySandeep replied:
You can take it that way, but I was expecting one method to test one behavior. So a single function "can have"  multiple tests. (Assume a  function has X and Y behaviours for a param then you'll write two tests for the function with different params the first one will test the X behaviour and the 2nd one will test the Y behaviour.)

You can find more details on writing a good unit test in [the wiki page](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests).

EDIT: You can also check these guidelines: https://github.com/oppia/oppia/wiki/Frontend-test-best-practices#best-practices-for-good-tests

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/topic-viewer-backend-api.service.spec.ts:
```diff
@@ -20,6 +20,9 @@
 // ^^^ This block is to be removed.
+import { TopicDataObjectFactory } from
+  'domain/topic_viewer/TopicDataObjectFactory';
+
```
Remove extra newline*

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/topic-viewer-backend-api.service.spec.ts:
```diff
@@ -20,6 +20,9 @@
 // the code corresponding to the spec is upgraded to Angular 8.
 import { UpgradedServices } from 'services/UpgradedServices';
 // ^^^ This block is to be removed.
+import { TopicDataObjectFactory } from
```
Are we expecting this to be above the comment on line 22?

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/topic-viewer-backend-api.service.ts:
```diff
@@ -20,20 +20,24 @@ require('domain/utilities/url-interpolation.service.ts');
+
 angular.module('oppia').factory('TopicViewerBackendApiService', [
   '$http', '$q', 'UrlInterpolationService', 'TOPIC_DATA_URL_TEMPLATE',
   function($http, $q, UrlInterpolationService, TOPIC_DATA_URL_TEMPLATE) {
```
The service should be imported in the angular module, right?

@bansalnitish, can you please confirm this?

bansalnitish replied:
What's this? I haven't followed the complete PR but we are not using an updated service like this in place of AngularJS service.

We add downgrade statement in Angular service and then use it using require statements itself. Let me know if you have some questions.

Vinit-Dantkale replied:
Oh its `require('domain/topic_viewer/TopicDataObjectFactory')` in service and `angular.module('oppia').factory(... , downgradeInjectable(...))` in Class file. Got it.

DubeySandeep commented on core/templates/dev/head/pages/topic-viewer-page/topic-viewer-page.controller.ts:
```diff
@@ -65,14 +65,14 @@ angular.module('oppia').directive('topicViewerPage', [
-                ctrl.degreesOfMastery = topicDataDict.degrees_of_mastery;
-                ctrl.skillDescriptions = topicDataDict.skill_descriptions;
-                ctrl.subtopics = topicDataDict.subtopics;
+              function(topicDataObject) {
```
Don't end the variable name for an object with `Object`, we only add Dicts at the end to explain the variable contains dict object.

Vinit-Dantkale replied:
I think only `topicData` sounds good

DubeySandeep commented on core/templates/dev/head/pages/topic-viewer-page/topic-viewer-page.controller.ts:
```diff
@@ -65,14 +65,14 @@ angular.module('oppia').directive('topicViewerPage', [
+                  topicDataObject.getCanonicalStories();
+                ctrl.degreesOfMastery = topicDataObject.getDegreesOfMastery();
+                ctrl.skillDescriptions = topicDataObject.getSkillDescription();
+                ctrl.subtopics = topicDataObject.getSubtopics();
```
I don't see any usage of additionalStory, uncategorizedSkillIds, skillDescriptions here, is this `TopicViewerBackendApiService` being used anywhere else? If not then I'm not sure why we are sending the data through API if it's not used anywhere? @aks681, Can you explain?

DubeySandeep replied:
@aks681, Can you please reply to this?

aks681 replied:
No, it's not used anywhere else.
Yeah, we don't use additionalStories now anywhere actually, but I think it's fine to leave that in since it will always be an empty list.
Not sure of uncategorizedSkillIds, but I thought we used skill descriptions in the practice sessions tab. Though I'd suggest not removing those fields, since they could be used in the future and are related to topics directly.

Vinit-Dantkale replied:
Hey @aks681 , can we get `node_count` value from any other class or can we assign it vague value? as I'm short of only 1 field here.
![image](https://user-images.githubusercontent.com/32016324/74030890-2a50f880-49d6-11ea-9142-51dc4b5d0762.png)

aks681 replied:
Oh, I was pretty sure the objects in canonical_story_dicts list had that field, can you check?

Vinit-Dantkale replied:
I did `console.log(response.data)` in `service.ts` file & didn't find it. Was my flow right?
![image](https://user-images.githubusercontent.com/32016324/74055746-9b100900-4a06-11ea-8b56-facf241a104b.png)

Vinit-Dantkale replied:
Following our discussions I made changes in backend

Vinit-Dantkale replied:
Following our discussions I made changes in backend.

DubeySandeep reviewed (commented):
@Vinit-Dantkale, I took a pass and left a few comments PTAL!

@aks681, I have got a few questions for you can you please check?

aks681 commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+      subtopics: any, degreesOfMastery: any, skillDescriptions: any) {
+    this._topicName = topicName;
+    this._topicId = topicId;
+    this._canonicalStories = canonicalStory;
```
canonicalStory => canonicalStories

aks681 commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+    this._topicName = topicName;
+    this._topicId = topicId;
+    this._canonicalStories = canonicalStory;
+    this._additionalStories = additionalStory;
```
Ditto

aks681 commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  getCanonicalStories(): any {
+    return this._canonicalStories;
```
Use `this._canonicalStories.slice()`, since just this, would return a pointer to the array potentially allowing the receiver to modify its contents. (Check TopicObjectFactory for examples)
Ditto elsewhere with arrays.

aks681 commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+  createFromBackendDict(topicDataDict: any): TopicData {
+    return new TopicData(
+      topicDataDict.topic_name, topicDataDict.topic_id,
+      topicDataDict.canonical_story_dicts,
```
No, don't pass dictionaries, use StorySummaryObjectFactory to create a list of StorySummary objects here and pass it (again, check TopicObjectFactory for examples). 
Though, you would need to add a `description` field to StorySummaryObjectFactory as well, since the dict here has that as well.

Vinit-Dantkale replied:
StorySummaryObjectFactory contains `node_count` and `story_is_published` fields also. Should I pass `null`?

Vinit-Dantkale replied:
If I change StorySummary then, do I have to change at places where StorySummary is being used?

Vinit-Dantkale replied:
Hey @aks681 , can you please tell me about this? Im kinda stuck with `node_count`

DubeySandeep replied:
I don't think that a good idea (but we can do so.). I think [this comment](https://github.com/oppia/oppia/pull/8476#discussion_r373779753) can help us to avoid this issue.

aks681 commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+    return new TopicData(
+      topicDataDict.topic_name, topicDataDict.topic_id,
+      topicDataDict.canonical_story_dicts,
+      topicDataDict.additional_story_dicts,
```
Ditto

aks681 commented on core/templates/dev/head/domain/topic_viewer/TopicDataObjectFactory.ts:
```diff
@@ -0,0 +1,86 @@
+      topicDataDict.topic_name, topicDataDict.topic_id,
+      topicDataDict.canonical_story_dicts,
+      topicDataDict.additional_story_dicts,
+      topicDataDict.uncategorized_skill_ids, topicDataDict.subtopics,
```
Ditto with subtopics, they are a list of objects in SubtopicObjectFactory.

Vinit-Dantkale replied:
Hey, I actually need a bit help on this as `subtopics` is array and `skill_descriptions` is a single object. The `create()` function in `SubtopicObjectFactory` takes `subtopic` dict and `skill_description`

aks681 reviewed (changes requested):
Left some comments, PTAL.

Commit:
Using objects for Subtopics and Uncategorized Skill

Commit:
Fixing Lint errors

Commit:
Fixing Lint errors

Commit:
Removed previous files

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating instances of frontend topic-data domain
```
Factory for creating instances of ReadOnlyTopic...*

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+
+import { Subtopic, SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+
```
Remove this line.

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+
```
Remove this line*

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+import { Subtopic, SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+
+import { SkillSummary, SkillSummaryObjectFactory } from
```
Add this in sorted order i.e, skill than Subtopic

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+    return this._topicId;
+  }
+
+  // Return type is any as there is no usage
```
I didn't get it? If it's of no use then don't add it to the domain object.

Vinit-Dantkale replied:
Right, I didn't convey it properly as per https://github.com/oppia/oppia/pull/8476#discussion_r374216609. Though I found `canonicalStoriesList` being used in `topic-viewer-page.directive.html`. So, thought better keep it rather than causing error.

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  // Return type is any as there is no usage
+  getAdditionalStories(): any {
```
Ditto*

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+    return this._subtopics.slice();
+  }
+
+  // Return type any as no appropriate class found for object construction
```
If that's the reason then please create an issue for creating a class for this kinda object or create a new class for the same. 

I tried looking into the backend and I think this can be represented as dict(string: number or None). Can you please check how we can represent this in typescript?

DubeySandeep replied:
https://stackoverflow.com/questions/42211175/typescript-hashmap-dictionary-interface
Maybe this would be helpful^
Basically, if we are sure that the degreesOfMastery is going to be a dict with key skillId: string and value degree: number | null, then let's not use `any`. Maybe we can define an interface and use it?

Vinit-Dantkale replied:
Thank You so much for suggesting this, I will try this.

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+
+  // Return Type any as no appropriate class found for object construction
+  getSkillDescriptions(): any {
+    return this._skillDescriptions;
```
Ditto I think this can be dict(string: string or None)

DubeySandeep commented on core/templates/dev/head/pages/topic-viewer-page/navbar-breadcrumb/topic-viewer-navbar-breadcrumb.directive.ts:
```diff
@@ -33,8 +33,8 @@ angular.module('oppia').directive('topicViewerNavbarBreadcrumb', [
               UrlService.getTopicNameFromLearnerUrl()).then(
-              function(topicDataDict) {
-                $scope.topicName = topicDataDict.topic_name;
+              function(topicData) {
```
readOnlyTopic*

DubeySandeep commented on core/templates/dev/head/pages/topic-viewer-page/topic-viewer-page.controller.ts:
```diff
@@ -65,14 +65,14 @@ angular.module('oppia').directive('topicViewerPage', [
-                ctrl.degreesOfMastery = topicDataDict.degrees_of_mastery;
-                ctrl.skillDescriptions = topicDataDict.skill_descriptions;
-                ctrl.subtopics = topicDataDict.subtopics;
+              function(topicData) {
```
readOnlyTopic*

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/topic-viewer-backend-api.service.ts:
```diff
@@ -17,23 +17,26 @@
 require('domain/utilities/url-interpolation.service.ts');
-
 require('domain/topic_viewer/topic-viewer-domain.constants.ajs.ts');
+require('domain/topic_viewer/ReadOnlyTopicObjectFactory');
```
Add this in sorted order

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactory.ts:
```diff
@@ -0,0 +1,128 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
Rename this file to read-only-topic-object.factory.ts and similarly change the test file name.

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactorySpec.ts:
```diff
@@ -0,0 +1,116 @@
+
+/**
+ * @fileoverview Tests for ReadOnlyTopicObjectFactory.
+   */
```
Delete the space in the beginning.

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/ReadOnlyTopicObjectFactorySpec.ts:
```diff
@@ -0,0 +1,116 @@
+  'domain/topic_viewer/ReadOnlyTopicObjectFactory';
+
+import { SubtopicObjectFactory } from 'domain/topic/SubtopicObjectFactory';
+
```
Delete unwanted newlines. (Check the guidelines [here](https://github.com/oppia/oppia/wiki/Coding-style-guide#webpack))

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/topic-viewer-backend-api.service.spec.ts:
```diff
@@ -19,6 +19,11 @@
+import { SkillSummaryObjectFactory } from
+  'domain/skill/SkillSummaryObjectFactory';
+import { SubtopicObjectFactory } from 'domain/topic/SubtopicObjectFactory';
+import { ReadOnlyTopicObjectFactory } from
```
Add this in sorted order

DubeySandeep commented on core/templates/dev/head/domain/topic_viewer/topic-viewer-backend-api.service.ts:
```diff
@@ -17,23 +17,26 @@
-    var topicDataDict = null;
+  '$http', '$q', 'ReadOnlyTopicObjectFactory', 'UrlInterpolationService',
+  'TOPIC_DATA_URL_TEMPLATE',
+  function($http, $q, ReadOnlyTopicObjectFactory, UrlInterpolationService,
```
```
'$http', '$q', 'ReadOnlyTopicObjectFactory', 'UrlInterpolationService',
  TOPIC_DATA_URL_TEMPLATE, function(
    $http, $q, ReadOnlyTopicObjectFactory, UrlInterpolationService,
    TOPIC_DATA_URL_TEMPLATE)```

(Basically break the line after parenthesis if the params are in multi-lines)

DubeySandeep reviewed (commented):
@Vinit-Dantkale, This PR looks way better now, I've left a few comments PTAL! I think it can be merged once these comments are resolved. Thanks for the good work! :)

Commit:
Changes in filename, spaces, sorting

Commit:
Only node_count remaining

kevinlee12 commented:
@Vinit-Dantkale is this ready for review? If so, please ping the relevant reviewers.

Vinit-Dantkale commented:
> @Vinit-Dantkale is this ready for review? If so, please ping the relevant reviewers.

Some last minor changes are remaining, I'll ping for sure

Commit:
Using StorySummary

Commit:
Fixing Linting Errors

Commit:
Fix Lint Error

Commit:
Fix Lints

Commit:
Added space for fileOverview

Commit:
Fixing Failing Tests

Vinit-Dantkale commented:
Can you PHAL @kevinlee12 , @DubeySandeep , @aks681 .

bansalnitish commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+    this._skillSummaryObjectFactory = skillSummaryObjectFactory;
+  }
+
+  getStorySummaryArray(storyDicts: any): Array<StorySummary> {
```
Is this really any?

Vinit-Dantkale replied:
Actually, it had one of the fields missing in dict. So, I didn't find relevant exact type.

bansalnitish commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+  getStorySummaryArray(storyDicts: any): Array<StorySummary> {
+    var storyReferenceObjectFactory = new StoryReferenceObjectFactory();
+    let storySummaryArray: Array<StorySummary> =
+    storyDicts.map((storyDict: any) => {
```
Ditto here

bansalnitish commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+    return storySummaryArray;
+  }
+
+  createFromBackendDict(topicDataDict: any): ReadOnlyTopic {
```
Ditto

Vinit-Dantkale replied:
This is actual dict used to created object, can't find a better type. I think I should add comment to remove the `any` ?

bansalnitish reviewed (commented):
Few nits, PTAL!

aks681 commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+export class ReadOnlyTopic {
+  _topicName: string;
+  _topicId: string;
+  _canonicalStories: Array<StorySummary>;
```
_canonicalStorySummaries.

aks681 commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+  _topicName: string;
+  _topicId: string;
+  _canonicalStories: Array<StorySummary>;
+  _additionalStories: Array<StorySummary>;
```
_additionalStorySummaries

aks681 commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+  _topicId: string;
+  _canonicalStories: Array<StorySummary>;
+  _additionalStories: Array<StorySummary>;
+  _uncategorizedSkills: Array<SkillSummary>;
```
_uncategorizedSkillSummaries

aks681 commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,126 @@
+    expect(_sampleReadOnlyTopic.getSubtopics()[0]._skillSummaries[0]._id).
+      toEqual('skill_id_2');
+    expect(_sampleReadOnlyTopic.getSubtopics()[0]._skillSummaries[0].
+      _description).toEqual('Skill Description 2');
```
use getId() and getDescription() on the skillSummary objects.

aks681 commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+    var storyReferenceObjectFactory = new StoryReferenceObjectFactory();
+    let storySummaryArray: Array<StorySummary> =
+    storyDicts.map((storyDict: any) => {
+      return new StorySummary( storyDict.id, storyDict.title,
```
Remove the space after bracket.

aks681 commented on core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,155 @@
+    storyDicts.map((storyDict: any) => {
+      return new StorySummary( storyDict.id, storyDict.title,
+        storyDict.node_count, storyDict.description,
+        storyReferenceObjectFactory.createFromStoryId(storyDict.id).
```
createFromStoryId() function always sets the published status to false, but here the published status is true, I think it should be fine to just pass 'true' here since in topic_viewer, we only return published stories from the backend.

A better alternative would be to modify topic_viewer.py and in the story summary dicts, returned from there, manually add `summary['published'] = true`, since only published stories are fetched there, then use that here to make this generalized. No need to create a StoryReferenceObjectFactory object here.

Vinit-Dantkale replied:
Where are we supposed to add `summary['published'] = true` ?

aks681 replied:
Somewhere after this [line ](https://github.com/oppia/oppia/blob/b36ede61274c4b06011925d31511b548f00193ef/core/controllers/topic_viewer.py#L80) in topic_viewer.py. Just iterate through all the stories and add the published = True, since earlier in the function, we only fetched the published stories, so it is fine to do this.
Then, you can use `storyDict.published` here directly.

Vinit-Dantkale replied:
![image](https://user-images.githubusercontent.com/32016324/74462997-1fef9c80-4eb7-11ea-8bc7-daa832ebe7f5.png)
I think you were talking about something like this

Vinit-Dantkale replied:
I did the changes, though `backend_tests` is not giving me 100% coverage. Says missing on 80-82 in `topic_viewer.py`. Should I remove `summary['published'] = True` for additional_story_dicts?

aks681 replied:
No, just modify topic_viewer_test, so you test those lines as well.

Vinit-Dantkale replied:
Got delayed because of midsem. Actually so far wherever I've printed `additional_story_dicts`, its been an empty array. Also, `expected_dict` has it empty.

aks681 reviewed (commented):
Looks much better now! Just a couple of comments, should be ready after this, thanks!

Commit:
Changes for StorySummary

Commit:
Fixing Lints

Commit:
fetched for resolving conflicts

Commit:
Merge conflicts and coverage

Commit:
fixing lints

Commit:
fix lint

aks681 reviewed (changes requested):
You have to change the story summary tile directive to use the 'get' functions of the story summary object. Since, when I load topic viewer, I get these errors:
![image](https://user-images.githubusercontent.com/22347970/75355590-c0d95100-58d4-11ea-8541-ff003960f57d.png)

Which is due to story viewer link not being parsed properly in the story summary tile directive.

Commit:
jumping to develop

Commit:
Fixing v1

Commit:
Akshay changes

Commit:
Merging from develop
fixed the errors in console and used --amend
Changes in subtopics-list dir.html
Fixing subtopics part

Vinit-Dantkale commented:
I have solved this error locally (race condition still exists). Wanted to know how should I check for skills_description and subtopics manually on webpage.

Commit:
Merge branch 'develop' into domain-object
Merge from develop
Sorted order in UpgradedServices.ts

DubeySandeep commented:
@Vinit-Dantkale Please merge your PR with upstream/develop to fix the flaky e2e test issue on CircleCI (if it's a flaky error).

kevinlee12 commented:
@Vinit-Dantkale have you pushed your merge from develop? We don't see it here.

kevinlee12 commented:
@bansalnitish ptal!

Vinit-Dantkale commented:
I thought Travis handles test fails caused by some other branch. I'll push though by merging upstream/develop.

kevinlee12 commented:
I think you're okay for now, no need to push unless you need to make a code change :)

bansalnitish commented on core/templates/services/UpgradedServices.ts:
```diff
@@ -135,6 +135,8 @@ import { PlaythroughIssueObjectFactory } from
 import { PlaythroughObjectFactory } from
   'domain/statistics/PlaythroughObjectFactory';
 import { PythonProgramTokenizer } from 'classifiers/python-program.tokenizer';
+import { ReadOnlyTopicObjectFactory } from
```
Add it below the ReadOnlyStoryNodeObjectFactory. We want to keep all the imports in sorted order.

Vinit-Dantkale replied:
Sure

bansalnitish reviewed (approved):
Minor nits, LGTM otherwise!

Vinit-Dantkale commented:
I needed some help with this merge conflict, Im not able to find `train_tab_should_be_displayed`.How should I handle this?
![image](https://user-images.githubusercontent.com/32016324/76188871-3f03f480-61ff-11ea-8cc3-c3481a9233e5.png)

I also feel this [file](https://github.com/oppia/oppia/blob/develop/core/controllers/topic_viewer.py) is also changed, but I can't find it in merge conflicts.

bansalnitish commented:
Hi @Vinit-Dantkale, `<<<HEAD` to `===` denotes your changes and `===` to `>>> develop` the code that is in develop. So looking at both pieces of the code you'll decide what code to keep (may be all) and what to remove.

Not all files will result in merge conflicts. If the github is able to resolve changes for you, it'll do for you, else it'll ask you. You can read online as well for merge conflicts, there is ton of information available.

Please let me know if you need any other help. Thanks!

Vinit-Dantkale commented:
Actually, I trying to say I didn't find this field in sampleDataResult in `topicViewerBackendApiService` spec file. Seems like this was added recently. But, I'll look into it, no problem. :)

Commit:
Merging from develop
Made changes for train_tab_should_be_displayed

bansalnitish commented:
Ah nice to see, you've been able to resolve it :). 

@aks681 could you review this please?

Commit:
Merge branch 'develop' into domain-object
Merged from develop

kevinlee12 commented:
@aks681 ptal!

aks681 reviewed (approved):
Lgtm as codeowner, thanks!

aks681 commented:
Note merge conflict

Commit:
Merge branch 'develop' into domain-object
Merge from upstream/develop

DubeySandeep commented on core/controllers/topic_viewer.py:
```diff
@@ -71,9 +71,11 @@ def get(self, topic_name):
-            in canonical_story_summaries]
+        canonical_story_dicts = []
+        for summary in canonical_story_summaries:
+            summary = summary.to_human_readable_dict()
```
story_summary_human_readable_dict OR story_summary_dict* (`summary` should represent a summary object)

Vinit-Dantkale replied:
Yeah, story_summary_dict sounds specific.

DubeySandeep commented on core/controllers/topic_viewer.py:
```diff
@@ -71,9 +71,11 @@ def get(self, topic_name):
+        for summary in canonical_story_summaries:
+            summary = summary.to_human_readable_dict()
+            summary['published'] = True
+            canonical_story_dicts.append(summary)
```
Do we need to do similar changes for `additional_story_dicts`?

Vinit-Dantkale replied:
Yeah

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,131 @@
+  'domain/skill/SkillSummaryObjectFactory';
+import { SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+describe('Topic Data Object Factory', () => {
```
Add a newline above*

Vinit-Dantkale replied:
added

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,159 @@
+@Injectable({
+  providedIn: 'root'
+})
+
```
Do we need a newline here?

Vinit-Dantkale replied:
no, not needed

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,159 @@
+    this._skillSummaryObjectFactory = skillSummaryObjectFactory;
+  }
+
+  getStorySummaryArray(storyDicts: any): Array<StorySummary> {
```
Why do we need this method? I don't see this as a part of `ReadOnlyTopicObjectFactory`. Can you explain why/where we need this?

Vinit-Dantkale replied:
Actually, the `canoncialStories` list and `additionalStories` have objects of type `StorySummary`.  So to avoid repetition of code while converting the dicts to object, I had made that function.

Vinit-Dantkale replied:
yeah, not a good choice. Understood where I got wrong.

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,159 @@
+  }
+
+  createFromBackendDict(topicDataDict: any): ReadOnlyTopic {
+    let subtopics: Array<Subtopic> =
```
break line with parenthesis*

Vinit-Dantkale replied:
` let subtopics: Array<Subtopic> = topicDataDict.subtopics.map( ` ?

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,159 @@
+
+  createFromBackendDict(topicDataDict: any): ReadOnlyTopic {
+    let subtopics: Array<Subtopic> =
+    topicDataDict.subtopics.map((subtopic: Subtopic) => {
```
is this need to be indented by 2?

Vinit-Dantkale replied:
Indent of 2 more space needed

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,159 @@
+      return this._subtopicObjectFactory.create(
+        subtopic, topicDataDict.skill_descriptions);
+    });
+    let uncategorizedSkills: Array<SkillSummary> =
```
Ditto as above^

Vinit-Dantkale replied:
ok

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,159 @@
+    });
+    let degreesOfMastery: IDegreesOfMastery = topicDataDict.degrees_of_mastery;
+    let skillDescriptions: ISkillDescriptions =
+      topicDataDict.skill_descriptions;
```
ditto here and elsewhere*

Vinit-Dantkale replied:
yes

DubeySandeep commented on core/templates/pages/topic-viewer-page/topic-viewer-page.controller.ts:
```diff
@@ -65,16 +66,16 @@ angular.module('oppia').directive('topicViewerPage', [
-                ctrl.subtopics = topicDataDict.subtopics;
+              function(readOnlyTopic) {
+                ctrl.topicId = readOnlyTopic.getTopicId();
+                ctrl.canonicalStoriesList =
```
```
... = (
    ...);
```

(Use parentheses to break the line.)

Vinit-Dantkale replied:
yes

DubeySandeep commented on core/templates/pages/topic-viewer-page/topic-viewer-page.controller.ts:
```diff
@@ -65,16 +66,16 @@ angular.module('oppia').directive('topicViewerPage', [
-                ctrl.topicId = topicDataDict.id;
-                ctrl.trainTabShouldBeDisplayed = (
-                  topicDataDict.train_tab_should_be_displayed);
+                ctrl.trainTabShouldBeDisplayed =
```
Use parenthesis, (like it was earlier)

Vinit-Dantkale replied:
yes

DubeySandeep commented on core/templates/pages/topic-viewer-page/topic-viewer-page.mainpage.html:
```diff
@@ -11,7 +11,6 @@
           <topic-viewer-navbar-breadcrumb>
           </topic-viewer-navbar-breadcrumb>
         </navbar-breadcrumb>
-
```
Undo this change

Vinit-Dantkale replied:
Undone

DubeySandeep reviewed (changes requested):
@Vinit-Dantkale, I've left a few minor comments PTAL!

kevinlee12 commented:
@Vinit-Dantkale please merge from develop!

Commit:
minor changes

Commit:
Merge branch 'develop' into domain-object
"Minor changes as per lint"

Commit:
Proper indentation

Commit:
Fixing merge conflict

Commit:
* Fixing Merge conflict

* Requested changes

* Adding TODO comment

kevintab95 reviewed (approved):
Code-owner files LGTM, thanks @Vinit-Dantkale!

kevintab95 commented:
Hi @Vinit-Dantkale, once you have addressed all the comments, please assign the reviewer to take another pass!

And @-mention them in the comments with a "PTAL". Thanks!

Vinit-Dantkale commented:
Yeah @kevintab95. Thank You for approval !!

Vinit-Dantkale commented:
@DubeySandeep ptal, thanks.

DubeySandeep commented on core/controllers/topic_viewer.py:
```diff
@@ -71,12 +71,14 @@ def get(self, topic_name):
 
         additional_story_dicts = [
-            summary.to_human_readable_dict() for summary
+            story_summary_dict.to_human_readable_dict() for story_summary_dict
```
Do we need to have the 'published' key in all the story dict?

Vinit-Dantkale replied:
@DubeySandeep , [here](https://github.com/oppia/oppia/pull/8476/#discussion_r378975859)
I feel this got missed out. The expected_dict is empty in topic_viewier_test.py. But will still add it incase if it changes in future.

Vinit-Dantkale replied:
@DubeySandeep , I tried this but coverage does not go to 100%. So, reverting to previous way.

DubeySandeep replied:
@Vinit-Dantkale, you can't leave this as is because of no 100% coverage. Try writing test to make coverage 100%

DubeySandeep replied:
Also, why the variable name is renamed?

Vinit-Dantkale replied:
For variable name, https://github.com/oppia/oppia/pull/8476#discussion_r395861169

Vinit-Dantkale replied:
I wrote a test but still, coverage was 95%. I'll try again.

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+/**
+ * @fileoverview Tests for read-only-topic-object.factory.ts
+ */
+import { TestBed } from '@angular/core/testing';
```
Add a newline above

Vinit-Dantkale replied:
ok

DubeySandeep replied:
is this resolve?

Vinit-Dantkale replied:
Yes.
![image](https://user-images.githubusercontent.com/32016324/78501214-60152200-7778-11ea-8bc0-12d85c73ffe9.png)

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+        title: 'Story Title',
+        description: 'Story Description',
+        node_count: 1,
+        published: true
```
I think we need a published key in the additional story as well, I've left a comment above for the same*

Vinit-Dantkale replied:
Yep, will look into it

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+    expect(_sampleReadOnlyTopic.getUncategorizedSkillsSummaries()[0].getId())
+      .toEqual('skill_id_1');
+    expect(_sampleReadOnlyTopic.getUncategorizedSkillsSummaries()[0].
+      getDescription()).toEqual('Skill Description 1');
```
`.getDescription` (break before period, like the way you have done on line 82)

DubeySandeep replied:
is this resolved? (Can you please reply to all the comments below?)

Vinit-Dantkale replied:
Yes it's resolved. I'll address it.

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+      'subtopic_name');
+    expect(_sampleReadOnlyTopic.getSubtopics()[0]._skillSummaries[0].getId()).
+      toEqual('skill_id_2');
+    expect(_sampleReadOnlyTopic.getSubtopics()[0]._skillSummaries[0].
```
Ditto here and elsewher*

Vinit-Dantkale replied:
okay

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+    expect(_sampleReadOnlyTopic.getTrainTabShouldBeDisplayed()).toEqual(false);
+  });
+
+  it('should check value of uncategorized skill object', () => {
```
skill summary objects, right? (skill object and skill summaries object are different)

Vinit-Dantkale replied:
sure, skill summary objects.

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+      getDescription()).toEqual('Skill Description 2');
+  });
+
+  it('should check type and values of Skill Description', () => {
```
instead of `check` I would suggest `return correct ...` (here and elsewhere)

Vinit-Dantkale replied:
ok

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+
+  it('should check the values of Canonical Stories', () => {
+    expect(_sampleReadOnlyTopic.getCanonicalStorySummaries()[0].getId()).
+      toEqual('0');
```
break before `.` (ditto below)

Vinit-Dantkale replied:
okay

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+    });
+  });
+
+  it('should check the values of Canonical Stories', () => {
```
`canonical stories` or `conanicalStories` (I will prefer the first one as the sting to be a human-readable text.)

Vinit-Dantkale replied:
okay

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+      toEqual(1);
+  });
+
+  it('should check the values of Additional Stories', () => {
```
Ditto as above (additional stories)

Vinit-Dantkale replied:
yeah

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.spec.ts:
```diff
@@ -0,0 +1,132 @@
+      getNodeCount()).toEqual(1);
+  });
+
+  it('should get the values as expected', () => {
```
value of what? 

Also, `should return the correct value ...`

Vinit-Dantkale replied:
Sounds good

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,154 @@
+  _skillDescriptions: ISkillDescriptions;
+  _trainTabShouldBeDisplayed: boolean;
+
+  constructor(topicName: string, topicId: string,
```
break line after parenthesis (The way you have written on line 115)

Vinit-Dantkale replied:
yes

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  getDegreesOfMastery(): IDegreesOfMastery {
+    return this._degreesOfMastery;
```
Do we need to return a coy of this dict as well? (not sure, can you please check?)

Vinit-Dantkale replied:
Yes, its needed when you include skills.

DubeySandeep commented on core/templates/domain/topic_viewer/read-only-topic-object.factory.ts:
```diff
@@ -0,0 +1,154 @@
+    this._skillSummaryObjectFactory = skillSummaryObjectFactory;
+  }
+
+  createFromBackendDict(topicDataDict: any): ReadOnlyTopic {
```
Add a TODO comment to replace `any` from here. 

(See other similar places for ref.)

Vinit-Dantkale replied:
Okay.
@DubeySandeep, For topic_viewer, I didn't find in issue #7176, #7165.

Vinit-Dantkale replied:
Okay

DubeySandeep commented on core/templates/domain/topic_viewer/topic-viewer-backend-api.service.spec.ts:
```diff
@@ -15,26 +15,52 @@
 /**
  * @fileoverview Unit tests for TopicViewerBackendApiService.
  */
-
```
Undo this change!

Vinit-Dantkale replied:
ok

DubeySandeep commented on core/templates/pages/topic-viewer-page/subtopics-list/subtopics-list.directive.html:
```diff
@@ -18,9 +18,9 @@
-        <subtopic-summary-tile skill-count="subtopic.skill_ids.length"
-                               subtopic-id="subtopic.id"
-                               subtopic-title="subtopic.title"
+        <subtopic-summary-tile skill-count="subtopic.getSkillSummaries().length"
```
getSkillSummaries will return skill_ids? (Just wanted to double-check)

Vinit-Dantkale replied:
Yes, as per SubtopicObjectFactory.ts

DubeySandeep commented on core/templates/pages/topic-viewer-page/topic-viewer-page.controller.ts:
```diff
@@ -21,12 +21,12 @@ require(
   'components/common-layout-directives/common-elements/' +
   'background-banner.directive.ts');
 require('components/skills-mastery-list/skills-mastery-list.directive.ts');
+require('domain/topic_viewer/topic-viewer-backend-api.service.ts');
```
Undo this change, (As services are imported after the directive, you can add a newline after line 29)

Vinit-Dantkale replied:
ok

DubeySandeep reviewed (approved):
@Vinit-Dantkale, This PR looks great, I have left a few minor comments (nit pick) and a few questions PTAL!

(This PR can be merged once you resolve all the comments.)

Commit:
Merge branch 'develop' into domain-object
Merging from develop

Commit:
full test coverage

kevinlee12 commented:
@Vinit-Dantkale please do not force push. We will close PRs that do so, see #5 in our [guide](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#setting-things-up). I'll close this PR, please reopen a new PR. thanks!
